### DISCUSSION TITLE
Use external shared library API

### DISCUSSION THREAD

**Discussion by musaleh17** [strongswan/strongswan#1510] [Q&A]
I want to use API from an external shared library located on my device, is there any guide describing the steps of adding external lib to the compilation mechanism for StrongSwan? 

- (.so, .a, .pc) already exist on my device just need to add it to the Strongswan makefile.

**Comment by strongX509** [comment]
What type of external shared library do you want to use? A third party crypto library?

**Reply by musaleh17** [reply]
Yes. Crypto lib.

**Comment by strongX509** [comment]
You have to write a new libstrongswan  plugin acting as a wrapper around your crypto library. Have a look at the following existing libstrongswan crypto plugins:

- [openssl](https://github.com/strongswan/strongswan/tree/master/src/libstrongswan/plugins/openssl)
- [wolfssl](https://github.com/strongswan/strongswan/tree/master/src/libstrongswan/plugins/wolfssl)
- [botan](https://github.com/strongswan/strongswan/tree/master/src/libstrongswan/plugins/botan) 
- [gcrypt](https://github.com/strongswan/strongswan/tree/master/src/libstrongswan/plugins/gcrypt)

**Reply by musaleh17** [reply]
Hi,
Our lib would replace the kernel IPSEC, not only crypto lib. 
I guess the examples above are used as libstrongswan plugins, but in my case, I want libcharon plugin example. 
are you recommending on specific plugin to follow?

**Reply by tobiasbrunner** [reply]
> Our lib would replace the kernel IPSEC ... but in my case, I want libcharon plugin example

And with that you aren't able to find such a plugin yourself in our codebase?

**Reply by musaleh17** [reply]
No, I see there are many plugin's, I asked if you recommend on someone of them...
